database/publicapikeys.go

```diff
@@ -0,0 +1,197 @@
+package database
+
+import (
+	"context"
+	"encoding/base64"
+	"time"
+
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/NebulousLabs/fastrand"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+type (
+	// TODO: I am still not sure whether we should use separate collections or
```

Comment:
Tbh after reviewing the PR it seems like using a separate collection results in a lot of blowup. e.g. double the amount of deletion, fetching etc. methods. More endpoints and so on.

When actually a public api key is just a normal api key with a skylink whitelist. So I think we could safe a lot of code if we just treated them the same as regular api keys but if the skylinks field is not empty we restrict access to the specified links.

Reply:
Right, this is one of the big questions here. I'm not completely sure about this because public and private API keys need to be treated in different ways in many situations.

Examples:
1. When listing/getting API keys, private API keys must not be displayed, while public ones should be displayed in their entirety.
2. Editing private API keys is not possible.

Otherwise, keeping both types of API keys in the same collection will make it a little more straightforward to validate them.

Also, we will need a special flag, differentiating between public and private API keys - the list of skylinks is not enough. A user can `PUT`/`PATCH` a public API key and remove all of the skylinks it enables and that shouldn't turn the public API key into a private one. This is a minor details but it's relevant to your comment, so I'm mentioning it right away.